apps/settings/js/modules/base/task_scheduler.js

```diff
@@ -0,0 +1,133 @@
+/**
+ * TaskScheduler helps manage tasks and ensures they are executed in
+ * sequential order. When a task of a certain type is enqueued, all pending
+ * tasks of the same type in the queue are removed. This avoids redundant
+ * queries and improves user perceived performance.
+ *
+ * @module Task_sheduler
+ */
+define(function() {
+  'use strict';
+
+  var Defer = require('modules/defer');
+  var Module = require('modules/base/module');
+  var Observable = require('modules/mvvm/observable');
+  var TaskSchedulerConstant = require('modules/base/task_scheduler_constant');
+  
+  /**
+   * @class TaskScheduler
+   * @returns {TaskScheduler}
+   */
+  var TaskScheduler = Module.create(function() {
+    this.super(Observable).call(this);
+    this.isLocked = false;
+    this._tasks = [];
+  }).extend(Observable);
+
+  Observable.defineObservableProperty(TaskScheduler.prototype, 'isLocked', {
+    value: false
+  });
+
+  /**
+   * We will keep some constants here at its prototype chain
+   *
+   * @access public
+   * @memberOf TaskScheduler
+   */
+  TaskScheduler.prototype.TASK_TYPES = TaskSchedulerConstant.TASK_TYPES;
```

Comment:
For a enumeration we usually use singular form (TASK_TYPE). The other thing is, `TASK_TYPE` is like a static constant. I would prefer to define it on `TaskScheduler` instead of `TaskScheduler.prototype`.